test/integration/stats_integration_test.cc

```diff
@@ -129,5 +162,34 @@ TEST_P(StatsIntegrationTest, WithTagSpecifierWithFixedValue) {
   EXPECT_EQ(live->tags()[0].value_, "xxx");
 }
 
+INSTANTIATE_TEST_SUITE_P(IpVersions, ClusterMemoryUtilization,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// TODO(cmluciano) Refactor once envoyproxy/envoy#5624 is solved
+// TODO(cmluciano) Add options to measure multiple workers & without stats
+TEST_P(ClusterMemoryUtilization, MemoryLargeClusterSizeWithStats) {
+
+  const size_t start_mem = Memory::Stats::totalCurrentlyAllocated() / 1000;
+  if (start_mem == 0) {
+    // Skip this test for platforms where we can't measure memory.
+    return;
+  }
+
+  auto IpVersions = TestBaseWithParam<Network::Address::IpVersion>::GetParam();
+
+  auto t1 = std::make_unique<BaseIntegrationTest>(IpVersions);
+  t1->initialize();
+  size_t m1 = Memory::Stats::totalCurrentlyAllocated();
+  EXPECT_LT(start_mem, m1 / 1000);
+  EXPECT_LT(m1 / 1000, 3500); // actual value: 3427 as of Feb 19, 2019
+  t1.reset(nullptr);
+
+  size_t m1001 = memoryConsumedWithClusters(1001, true);
+  EXPECT_LT(start_mem / 1000, m1001 / 1000);
+  size_t m_per_cluster = (m1001 - m1) / 1000;
+  EXPECT_LT(m_per_cluster, 57900); // actual value: 57872 as of Feb 19, 2019
```

Comment:
Is his true on all build types?

Reply:
In general these memory-tests have been written with the assumption that only tcmalloc has accurate memory-measuring technology, and so across all build types that use tcmalloc we get the same numbers here exactly (assuming 64-bit of course).

Reply:
What about libc++? See #5483 for context.

Reply:
@PiotrSikora Islibstdc++ vs libc++ a dev-controllable option? If so, then I agree that absolute memory values will be affected, and this test will either have to be sloppy enough or to special-case alternative values.

Reply:
Yes, we've been supporting libc++ for a while now using `--config libc++`, and it's tested in CI as part of `bazel.compile_time_options`, so presumably this wouldn't pass CI anyway.

Reply:
I changed to EXPECT_EQ and ran with both libc++ and clang

```
libc++ 1 cluster
4250-Expected equality of these values:
4251-  m1 / 1000
4252-    Which is: 3134
4253:  3500
4254-[2019-02-26 16:41:25.950][15][info][testing] [test/integration/server.cc:180] stopping integration test server
4255-[2019-02-26 16:41:25.952][15][debug][client] [source/common/http/codec_client.cc:26] [C28] connecting
4256-[2019-02-26 16:41:25.952][15][debug][connection] [source/common/network/connection_impl.cc:639] [C28] connecting to 127.0.0.1:33969
--
libc++ 1 cluster
20826-Expected equality of these values:
20827-  m1 / 1000
20828-    Which is: 3134
20829:  3500
20830-[2019-02-26 16:41:32.788][15][info][testing] [test/integration/server.cc:180] stopping integration test server
20831-[2019-02-26 16:41:32.789][15][debug][client] [source/common/http/codec_client.cc:26] [C32] connecting
20832-[2019-02-26 16:41:32.789][15][debug][connection] [source/common/network/connection_impl.cc:639] [C32] connecting to [::1]:44579

libc++ 1001 cluster
19537-Expected equality of these values:
19538-  m_per_cluster
19539-    Which is: 52249
19540:  57900
19541-[2019-02-26 16:41:32.529][15][info][testing] [test/integration/server.cc:180] stopping integration test server
19542-[2019-02-26 16:41:32.530][15][debug][client] [source/common/http/codec_client.cc:26] [C30] connecting
19543-[2019-02-26 16:41:32.530][15][debug][connection] [source/common/network/connection_impl.cc:639] [C30] connecting to 127.0.0.1:36595
--
libc++ 1001 cluster
36113-Expected equality of these values:
36114-  m_per_cluster
36115-    Which is: 52249
36116:  57900
36117-[2019-02-26 16:41:39.291][15][info][testing] [test/integration/server.cc:180] stopping integration test server
36118-[2019-02-26 16:41:39.292][15][debug][client] [source/common/http/codec_client.cc:26] [C34] connecting
36119-[2019-02-26 16:41:39.292][15][debug][connection] [source/common/network/connection_impl.cc:639] [C34] connecting to [::1]:40167
```

```
clang 1 cluster
4250-Expected equality of these values:
4251-  m1 / 1000
4252-    Which is: 3442
4253:  3500
4254-[2019-02-26 16:57:39.257][15][info][testing] [test/integration/server.cc:180] stopping integration test server
4255-[2019-02-26 16:57:39.258][15][debug][client] [source/common/http/codec_client.cc:26] [C28] connecting
4256-[2019-02-26 16:57:39.258][15][debug][connection] [source/common/network/connection_impl.cc:639] [C28] connecting to 127.0.0.1:38053
--
cluster 1 cluster
20826-Expected equality of these values:
20827-  m1 / 1000
20828-    Which is: 3442
20829:  3500
20830-[2019-02-26 16:57:43.112][15][info][testing] [test/integration/server.cc:180] stopping integration test server
20831-[2019-02-26 16:57:43.113][15][debug][client] [source/common/http/codec_client.cc:26] [C32] connecting
20832-[2019-02-26 16:57:43.113][15][debug][connection] [source/common/network/connection_impl.cc:639] [C32] connecting to [::1]:33279

clang++ 1001 cluster
19537-Expected equality of these values:
19538-  m_per_cluster
19539-    Which is: 56119
19540:  57900
19541-[2019-02-26 16:57:42.842][15][info][testing] [test/integration/server.cc:180] stopping integration test server
19542-[2019-02-26 16:57:42.843][15][debug][client] [source/common/http/codec_client.cc:26] [C30] connecting
19543-[2019-02-26 16:57:42.843][15][debug][connection] [source/common/network/connection_impl.cc:639] [C30] connecting to 127.0.0.1:39287
--
clang++ 1001 cluster
36113-Expected equality of these values:
36114-  m_per_cluster
36115-    Which is: 56119
36116:  57900
36117-[2019-02-26 16:57:46.683][15][info][testing] [test/integration/server.cc:180] stopping integration test server
36118-[2019-02-26 16:57:46.684][15][debug][client] [source/common/http/codec_client.cc:26] [C34] connecting
36119-[2019-02-26 16:57:46.684][15][debug][connection] [source/common/network/connection_impl.cc:639] [C34] connecting to [::1]:40181
```

Reply:
Unfortunately it looks like  the value of 57900 has already been violated and we're now at 57936 vs the 57872 on Jan 19th.